src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisResult.cs

```diff
@@ -13,195 +14,148 @@ namespace Microsoft.CodeAnalysis.Diagnostics
+            Analyzers = analyzers;
+            SyntaxDiagnostics = localSyntaxDiagnostics;
+            SemanticDiagnostics = localSemanticDiagnostics;
+            CompilationDiagnostics = nonLocalDiagnostics;
+            AnalyzerTelemetryInfo = analyzerTelemetryInfo;
         }
 
-        private static Dictionary<DiagnosticAnalyzer, TimeSpan> CreateAnalyzerExecutionTimeMap(ImmutableArray<DiagnosticAnalyzer> analyzers)
+        /// <summary>
+        /// Analyzers corresponding to this analysis result.
+        /// </summary>
+        public ImmutableArray<DiagnosticAnalyzer> Analyzers { get; private set; }
+
+        /// <summary>
+        /// Syntax diagnostics reported by the <see cref="Analyzers"/>.
+        /// </summary>
+        public ImmutableDictionary<SyntaxTree, ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>>> SyntaxDiagnostics { get; private set; }
+
+        /// <summary>
+        /// Semantic diagnostics reported by the <see cref="Analyzers"/>.
+        /// </summary>
+        public ImmutableDictionary<SyntaxTree, ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>>> SemanticDiagnostics { get; private set; }
+
+        /// <summary>
+        /// Compilation diagnostics reported by the <see cref="Analyzers"/>.
+        /// </summary>
+        public ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>> CompilationDiagnostics { get; private set; }
+
+        /// <summary>
+        /// Analyzer telemetry info (register action counts and execution times).
+        /// </summary>
+        public ImmutableDictionary<DiagnosticAnalyzer, AnalyzerTelemetryInfo> AnalyzerTelemetryInfo { get; private set; }
+
+        /// <summary>
+        /// Gets all the diagnostics reported by the given <paramref name="analyzer"/>.
+        /// </summary>
+        public ImmutableArray<Diagnostic> GetAllDiagnostics(DiagnosticAnalyzer analyzer)
         {
-            var map = new Dictionary<DiagnosticAnalyzer, TimeSpan>(analyzers.Length);
-            foreach (var analyzer in analyzers)
+            if (!Analyzers.Contains(analyzer))
             {
-                map[analyzer] = default(TimeSpan);
+                throw new ArgumentException(CodeAnalysisResources.UnsupportedAnalyzerInstance, nameof(analyzer));
             }
 
-            return map;
-        }
-
-        public TimeSpan GetAnalyzerExecutionTime(DiagnosticAnalyzer analyzer)
-        {
-            Debug.Assert(_analyzerExecutionTimeOpt != null);
-
-            lock (_gate)
+            if (Analyzers.Length == 1 && Analyzers[0] == analyzer)
             {
-                return _analyzerExecutionTimeOpt[analyzer];
+                return GetAllDiagnostics();
             }
+
+            return GetDiagnostics(SpecializedCollections.SingletonEnumerable(analyzer), getLocalSyntaxDiagnostics: true, getLocalSemanticDiagnostics: true, getNonLocalDiagnostics: true);
```

Comment:
This method is making two calls to `GetDiagnostics` essentially.  In one we specify a set of options, in the other we assume the defaults (indirictly through `GetallDiagnostics`).  I would prefer we are consistent here and specify the options in both places.  That way it's clear we are getting the same data in both cases.